File: packages/rust/examples/simple_math.rs
/*!
Simple Example: Math Chain Processing

Demonstrates modular chain processing with math links and hook.
Shows the new modular structure: core protocols, component implementations.
*/

use std::collections::HashMap;
use codeuchain::core::state::State;

// Import from local examples
mod components;
use components::{BasicChain, MathLink, LoggingHook};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Set up the chain using component implementations
    let mut chain = BasicChain::new();
    chain.add_link("sum".to_string(), Box::new(MathLink::new("sum".to_string())));
    chain.add_link("mean".to_string(), Box::new(MathLink::new("mean".to_string())));
    chain.connect(
        "sum".to_string(),
        "mean".to_string(),
        |ctx| ctx.get("result").is_some(),
    );
    chain.use_hook(Box::new(LoggingHook::new()));

    // Run with initial state
    let mut data = HashMap::new();
    data.insert("numbers".to_string(), serde_json::json!([1, 2, 3, 4, 5]));
    let ctx = State::new(data);

    let result = chain.run(ctx).await;
    let result = match result {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Error: {}", e);
            return Ok(());
        }
    };

    println!("Final result: {:?}", result.get("result"));
    println!("Full state: {:?}", result.to_hashmap());

    Ok(())
}

File: packages/python/examples/simple_math.py
"""
Simple Example: Math Chain Processing

Demonstrates modular chain processing with math links and middleware.
Shows the new modular structure: core protocols, component implementations.
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

import asyncio
from codeuchain.core import Context
from components.chains import BasicChain
from components.links import MathLink
from components.middleware import LoggingMiddleware


async def main():
    # Set up the chain using component implementations
    chain = BasicChain()
    chain.add_link("sum", MathLink("sum"))
    chain.add_link("mean", MathLink("mean"))
    chain.connect("sum", "mean", lambda ctx: ctx.get("result") is not None)
    chain.use_middleware(LoggingMiddleware())
    
    # Run with initial context
    ctx = Context({"numbers": [1, 2, 3, 4, 5]})
    result = await chain.run(ctx)
    
    print(f"Final result: {result.get('result')}")  # Mean: 3.0
    print(f"Full context: {result.to_dict()}")  # Shows all data


if __name__ == "__main__":
    asyncio.run(main())